crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1496,33 +1434,227 @@ impl AuthorityPerEpochStore {
+    }
+
+    pub(crate) async fn process_consensus_transactions_and_commit_boundary<
+        C: CheckpointServiceNotify,
+    >(
+        &self,
+        transactions: Vec<VerifiedSequencedConsensusTransaction>,
+        round: Round,
+        timestamp: u64,
+        checkpoint_service: &Arc<C>,
+        parent_sync_store: impl ParentSync,
+    ) -> SuiResult<Vec<VerifiedExecutableTransaction>> {
+        let mut batch = self.db_batch();
+        let (executable_txns, lock) = self
+            .process_consensus_transactions(
+                &mut batch,
+                transactions,
+                checkpoint_service,
+                parent_sync_store,
+            )
+            .await?;
+        batch.write()?;
+        drop(lock);
+        if let Some(checkpoint) = self.handle_commit_boundary(round, timestamp)? {
+            let final_checkpoint = checkpoint.details.last_of_epoch;
+            checkpoint_service.notify_checkpoint(self, checkpoint)?;
+            if final_checkpoint {
+                tracing::info!(epoch=?self.epoch(), "Received 2f+1 EndOfPublish messages, notifying last checkpoint");
+                self.record_end_of_message_quorum_time_metric();
+            }
+        }
+        Ok(executable_txns)
+    }
+
+    #[cfg(test)]
+    pub(crate) async fn process_consensus_transactions_for_tests<C: CheckpointServiceNotify>(
+        &self,
+        transactions: Vec<VerifiedSequencedConsensusTransaction>,
+        checkpoint_service: &Arc<C>,
+        parent_sync_store: impl ParentSync,
+    ) -> SuiResult<Vec<VerifiedExecutableTransaction>> {
+        let mut batch = self.db_batch();
+        let (certs, _lock) = self
+            .process_consensus_transactions(
+                &mut batch,
+                transactions,
+                checkpoint_service,
+                parent_sync_store,
+            )
+            .await?;
+        batch.write()?;
+        Ok(certs)
+    }
+
     /// Depending on the type of the VerifiedSequencedConsensusTransaction wrappers,
     /// - Verify and initialize the state to execute the certificates.
     ///   Return VerifiedCertificates for each executable certificate
     /// - Or update the state for checkpoint or epoch change protocol.
     pub(crate) async fn process_consensus_transactions<C: CheckpointServiceNotify>(
         &self,
-        transactions: Vec<VerifiedSequencedConsensusTransaction>,
+        batch: &mut DBBatch,
+        mut transactions: Vec<VerifiedSequencedConsensusTransaction>,
         checkpoint_service: &Arc<C>,
         parent_sync_store: impl ParentSync,
-    ) -> SuiResult<Vec<VerifiedExecutableTransaction>> {
+    ) -> SuiResult<(
+        Vec<VerifiedExecutableTransaction>,
+        Option<parking_lot::RwLockWriteGuard<ReconfigState>>,
+    )> {
         let mut verified_certificates = Vec::new();
-        for tx in transactions {
+
+        // sort all end of publish messages to the end of the list
+        transactions.sort_by(|a, b| {
```

Comment:
It feels like we don't really need sort here.
Only thing we need is making two iterators, one that filters end_of_publish transactions, and another that filters non-end_of_publish?

Reply:
that may produce borrow checker problems, but i'll give it a try

Reply:
since we were constructing the vector from scratch i instead just build two vectors and then chain their iters.

Reply:
This is no longer needed right? (handler already sorts them, and we can just change code there to assert protocol version)